packages/tools/src/utilities/contourROITool/Interpolation/freehandInterpolate/interpolate.ts

```diff
@@ -0,0 +1,656 @@
+  );
+}
+
+/**
+ * _editInterpolatedContour - Edits an interpolated polygon on the imageId
+ * that corresponds to the specified ROIContour.
+ *
+ * @param interpolated3DPoints - object, The polygon to add to the ROIContour.
+ * @param sliceIndex - Number, The slice index to interpolate.
+ * @param referencedToolData - type, The toolData of another polygon in the
+ * ROIContour, to assign appropriate metadata to the new polygon.
+ * @param eventData - object
+ * @returns null
+ */
+function _editInterpolatedContour(
+  interpolated3DPoints: { x: number[]; y: number[]; z: number[] },
+  sliceIndex,
+  referencedToolData,
+  eventData
+) {
+  const { viewport } = eventData;
+  const annotations = annotation.state.getAnnotations(
+    TOOL_NAMES.CONTOUR_ROI_TOOL,
+    viewport.element
+  );
+
+  // Find the index of the polygon on this slice corresponding to
+  // The ROIContour.
+  let toolDataIndex;
+
+  for (let i = 0; i < annotations.length; i++) {
+    const annotation = annotations[i] as InterpolationROIAnnotation;
+    if (
+      annotation.interpolationUID === referencedToolData.interpolationUID &&
+      annotation.metadata.referencedSliceIndex === sliceIndex
+    ) {
+      toolDataIndex = i;
+      break;
+    }
+  }
+
+  const oldToolData = annotations[toolDataIndex] as InterpolationROIAnnotation;
+  const points = [];
+
+  for (let i = 0; i < interpolated3DPoints.x.length; i++) {
+    points.push([
+      interpolated3DPoints.x[i],
+      interpolated3DPoints.y[i],
+      interpolated3DPoints.z[i],
+    ]);
+  }
+  const interpolatedAnnotation = getToolData(eventData, points, oldToolData);
+  interpolatedAnnotation.metadata.referencedImageId =
+    oldToolData.metadata.referencedImageId;
+  interpolatedAnnotation.metadata.referencedSliceIndex =
+    oldToolData.metadata.referencedSliceIndex;
+  // To update existing annotation, not intend to add or remove
+  interpolatedAnnotation.annotationUID = oldToolData.annotationUID;
+  // Skip triggering events on removal of interpolated roi's.
+  annotation.state.removeAnnotation(oldToolData.annotationUID, true);
+  annotation.state.addAnnotation(
+    interpolatedAnnotation,
+    viewport.element,
+    true
+  );
+}
+
+/**
+ * _generateInterpolatedOpenContour - Interpolate an open contour between c1ir
+ * and c2ir at the zInterp position.
+ *
+ * @param c1ir - object, The interpolated c1 contour with
+ *                                  superfluous nodes removed.
+ * @param c2ir - object, The interpolated c2 contour with
+ *                                  superfluous nodes removed.
+ * @param zInterp - Number, The z- coordinate of the desired
+ *                                  interpolation.
+ * @param c1HasMoreNodes - boolean, True if c1 has more original nodes
+ *                                  than c2.
+ * @returns object, The interpolated contour at z=zInterp.
+ */
+function _generateInterpolatedOpenContour(c1ir, c2ir, zInterp, c1HasMoreNodes) {
+  const cInterp = {
+    x: [],
```

Comment:
Consider using TypedArray values here instead - you are doing quite a lot of manipulation, and the typed arrays are noticeably faster for that sort of thing, not to mention more space efficient.